opacus/grad_sample/grad_sample_module.py

```diff
@@ -219,20 +220,22 @@ def add_hooks(
             self.autograd_grad_sample_hooks = self._module.autograd_grad_sample_hooks
 
         for _module_name, module in trainable_modules(self._module):
-            if type(module) in self.GRAD_SAMPLERS:
-                self.autograd_grad_sample_hooks.append(
-                    module.register_forward_hook(self.capture_activations_hook)
-                )
+            if not type(module) in self.GRAD_SAMPLERS: 
+                prepare_layer(module)
```

Comment:
question: can we call `prepare_layer` from within `compute_per_sample_gradient`? To avoid extra preparation step

Reply:
Oh, and we probably want to add extra bits to '_close()' method to remove extra attributes we add in `prepare_layer`

Reply:
> question: can we call `prepare_layer` from within `compute_per_sample_gradient`? To avoid extra preparation step

Sounds a bit wasteful no? Especially since GSM already does a preparation step to add hooks.